spec/upload_to_s3_spec.rb

```diff
@@ -0,0 +1,95 @@
+require 'tempfile'
+require_relative './spec_helper'
+
+describe Fastlane::Actions::UploadToS3Action do
+  let(:client) { instance_double(Aws::S3::Client) }
+  let(:test_bucket) { 'a8c-wpmrt-unit-tests-bucket' }
+
+  before do
+    allow(Aws::S3::Client).to receive(:new).and_return(client)
+  end
+
+  # Stub head_object to return a specific content_length
+  def stub_s3_head_request(key, content_length)
+    allow(client).to receive(:head_object)
+      .with(bucket: test_bucket, key: key)
+      .and_return(Aws::S3::Types::HeadObjectOutput.new(content_length: content_length))
+  end
+
+  # Allow us to do `.with` matching against a `File` instance to a particular path in RSpec expectations
+  # Because `File.open(path)` returns a different instance of `File` for the same path on each call)
+  RSpec::Matchers.define :file_instance_of do |path|
+    match { |actual| actual.is_a?(File) && actual.path == path }
+  end
```

Comment:
I wonder if it would be worth it to move this in `spec_helper.rb` for other, future tests to use… 🤷

Reply:
Moved in 8d543c0ad2d9e80c8989a0fdf28efd3207590f70, because I could see it being helpful for translation stuff too (really anything where we're pointing at a file).

Reply:
To be fair, for translation actions I did TDD, which led me to [create APIs that could equally take either a file path (`String`) or an `IO`/`File`](https://github.com/wordpress-mobile/release-toolkit/blob/trunk/lib/fastlane/plugin/wpmreleasetoolkit/helper/ios/ios_l10n_helper.rb#L127-L128) for versatility.

This allowed me to [use `StringIO` everywhere in my Unit Tests where I needed a pseudo-`File`](https://github.com/wordpress-mobile/release-toolkit/blob/trunk/spec/ios_l10n_helper_spec.rb#L183), which avoided me having to create actual temp files in the filesystem altogether, and instead have everything IO-related be in-memory [during all my tests (except the only test checking that passing a `String` worked, obviously)](https://github.com/wordpress-mobile/release-toolkit/blob/trunk/spec/ios_l10n_helper_spec.rb#L193-L195).

---

We could have considered going the same route as well for your `s3_upload_file` action, but that would have required some logic on the implementation's side (to convert the `String` file path to an `IO` via something like `file = File.open(file) if file.is_a?(String)` … but then that begs the question of how to close it only if it was opened by us (i.e. was a `String`) but not close it ourselves if the caller was the one who did the `open`… so probably not worth the added complexity in the implementation just to support passing an `IO`, which in the case of this action would only used in unit tests anyway but unlikely in real world call sites.

That was easier for my `ios_l10n_helper` case because the implementation already used `IO.copy_stream` which already supports either an `IO` or a `String` path so I was able to support that duality at zero cost here, unlike in your case for which it might then not be worth the effort, which is why I ended up keeping your action API and using the custom matcher on `File` instead.

Reply:
💡  An alternative idea (it's too late now, and not worth changing anymore but just for your information and Ruby culture, or a future refactoring) would have been to do use this in our unit tests:
```ruby
file_instance = StringIO.new
allow(File).to receive(:open).with(file_path).and_return(file_instance)
```
And then we would have been able to match using `.with(file_instance, …)` directly, without the need for the custom matcher above… 🤷  eh.